src/SmoFacade/Listener.cs

```diff
@@ -154,33 +153,47 @@ private static string AgListenerName(string dataSource)
     /// <param name="agReplicaInstanceName">The specific name of a replica instance within the AG</param>
     private static string ResolveDnsAndGetHostName(string agListenerDomain, string agReplicaInstanceName)
     {
-      Tuple<string, string> domain_port = SplitDomainPort(agListenerDomain);
-      if (Environment.OSVersion.Platform == PlatformID.Unix)
+      // sometimes instances/listners have ports and named instances. Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      (string listenerDomain, string listenerPort) = SplitDomainPort(agListenerDomain);
+      (string instanceDomain, string instancePort) = SplitDomainPort(agReplicaInstanceName);
+      // we want to add back instancePort if it exists, otherwise the server port
+      string portToUse = instancePort ?? listenerPort;
+      try
       {
-        string[] domainFragments = domain_port.Item1.Split('.');
-        domainFragments[0] = "";
-        agReplicaInstanceName += string.Join(".", domainFragments);
+        return $"{Dns.GetHostEntry(instanceDomain).HostName}{portToUse}";
       }
-      return Dns.GetHostEntry(agReplicaInstanceName).HostName+domain_port.Item2;
+      catch (System.Net.Sockets.SocketException e)
+      {
+        Console.WriteLine($"Unable to get dns for instance '{instanceDomain}'\n{e}");
+
+        // Re-try with full domain appended to the instanceDomain by stripping off domain from listner and appending to instance
+        // eg: if listner is "abc.def.ghi" we want to append ".def.ghi" to the instance
+        string[] domainFragments = listenerDomain.Split('.');
+        domainFragments[0] = "";
+        string instanceDomainFull = $"{instanceDomain}{string.Join(".", domainFragments)}";
 
+        Console.WriteLine($"Retrying with full domain '{instanceDomainFull}");

```

Comment:
Do we want these logging messages in the final release - I think they would be helpful for debugging for anyone using this

What do you guys think?

Reply:
Console logging isn't desireable in a Nuget package release IMO.
If you want log statements, they should go through Microsoft ILogger, for which there are many extensions to... like nlog
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0

Reply:
Had a look at `ILogger` and looks like it might be better to implement that (if we agree to do so) in it's own PR. There are many components that we will need to create for it to function (Logger models, categories, logging levels, etc.)

Shelfing for now, will come back to it if we feel it is necessary